// pinax-images-panel/lib/components/PreviewImage.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var PreviewImage = function (_React$Component) {
    _inherits(PreviewImage, _React$Component);

    function PreviewImage(props) {
        _classCallCheck(this, PreviewImage);

        var _this = _possibleConstructorReturn(this, Object.getPrototypeOf(PreviewImage).call(this, props));

        _this.handleDeleteClick = _this.handleDeleteClick.bind(_this);
        _this.handleMarkPrimaryClick = _this.handleMarkPrimaryClick.bind(_this);
        _this.render = _this.render.bind(_this);
        return _this;
    }

    _createClass(PreviewImage, [{
        key: 'handleDeleteClick',
        value: function handleDeleteClick() {
            this.props.deleteImage(this.props.image);
        }
    }, {
        key: 'handleMarkPrimaryClick',
        value: function handleMarkPrimaryClick() {
            this.props.markPrimaryImage(this.props.image);
        }
    }, {
        key: 'render',
        value: function render() {
            var previewImage = {};
            var selectedImageClassName = 'selected-image';
            if (this.props.image.is_primary) {
                selectedImageClassName += ' primary';
            }
            if (this.props.image.medium_thumbnail) {
                var thumb = this.props.image.medium_thumbnail;
                var background = 'url(' + thumb + ')';
                previewImage = {
                    backgroundImage: background
                };
            } else {
                selectedImageClassName += ' empty';
            }
            return _react2.default.createElement(
                'div',
                { className: 'col-md-9' },
                _react2.default.createElement(
                    'div',
                    { className: selectedImageClassName, style: previewImage },
                    _react2.default.createElement(
                        'div',
                        { id: 'overlay' },
                        _react2.default.createElement(
                            'a',
                            { href: '#', className: 'delete-image', onClick: this.handleDeleteClick },
                            _react2.default.createElement('i', { className: 'fa fa-trash fa-3x' })
                        ),
                        _react2.default.createElement(
                            'a',
                            { href: '#', className: 'make-primary-image', onClick: this.handleMarkPrimaryClick },
                            _react2.default.createElement('i', { className: 'fa fa-check fa-3x' })
                        )
                    )
                )
            );
        }
    }]);

    return PreviewImage;
}(_react2.default.Component);

PreviewImage.propTypes = {
    image: _react2.default.PropTypes.object.isRequired,
    deleteImage: _react2.default.PropTypes.func.isRequired,
    markPrimaryImage: _react2.default.PropTypes.func.isRequired
};

module.exports = PreviewImage;
